controllers/activemqartemis_controller.go

```diff
@@ -46,57 +46,118 @@ import (
+	} else {
+		m.effectiveHandler.Merge(incoming)
+	}
+}
+
+func (m *SecurityCRMerger) GetEffectiveHandler() common.ActiveMQArtemisConfigHandler {
+	return m.effectiveHandler
+}
+
+func GetSecurityConfigHandler(secNamespacedName types.NamespacedName) (handler common.ActiveMQArtemisConfigHandler) {
+	if handler, ok := namespaceToConfigHandler[secNamespacedName]; ok {
+		return handler
 	}
 	return nil
 }
 
-func (r *ActiveMQArtemisReconciler) UpdatePodForSecurity(securityHandlerNamespacedName types.NamespacedName, handler common.ActiveMQArtemisConfigHandler) error {
+func GetBrokerSecurityConfigHandler(brokerNamespacedName types.NamespacedName) (common.ActiveMQArtemisConfigHandler, map[string]string) {
+
+	merger := SecurityCRMerger{
+		ForBrokerCR:      brokerNamespacedName,
+		effectiveHandler: nil,
+	}
+	appliedSecurityMap := make(map[string]string)
+	for k, v := range namespaceToConfigHandler {
+		if v.IsApplicableFor(brokerNamespacedName) {
+			clog.V(1).Info("Got a security CR", "security", k, "for", brokerNamespacedName)
+			merger.Update(v)
+			secHandler := v.(*ActiveMQArtemisSecurityConfigHandler)
+			appliedSecurityMap[secHandler.SecurityCR.Name+"."+secHandler.SecurityCR.Namespace] = secHandler.SecurityCR.ResourceVersion
+		}
+	}
+	return merger.GetEffectiveHandler(), appliedSecurityMap
+}
+
+func (r *ActiveMQArtemisReconciler) UpdatePodForSecurity(securityHandlerNamespacedName types.NamespacedName, handler common.ActiveMQArtemisConfigHandler, isRemove bool) error {
 
 	existingCrs := &brokerv1beta1.ActiveMQArtemisList{}
 	var err error
 	opts := &rtclient.ListOptions{}
 	if err = r.Client.List(context.TODO(), existingCrs, opts); err == nil {
 		var candidate types.NamespacedName
-		for index, artemis := range existingCrs.Items {
+		for _, artemis := range existingCrs.Items {
 			candidate.Name = artemis.Name
 			candidate.Namespace = artemis.Namespace
 			if handler.IsApplicableFor(candidate) {
-				clog.Info("force reconcile for security", "handler", securityHandlerNamespacedName, "CR", candidate)
-				r.events <- event.GenericEvent{Object: &existingCrs.Items[index]}
+				// here check if the handler is already applied
+				// if this security cr already applied, don't reconcile
+				if !r.IsSecurityApplied(&artemis, handler) || isRemove {
+					clog.Info("force reconcile for security", "handler", securityHandlerNamespacedName, "CR", candidate)
+					r.events <- event.GenericEvent{Object: &artemis}
+				}
 			}
 		}
 	}
 	return err
 }
 
+func (r *ActiveMQArtemisReconciler) IsSecurityApplied(artemis *brokerv1beta1.ActiveMQArtemis, handler common.ActiveMQArtemisConfigHandler) bool {
+	//looking for the secret <cr-name>-applied-security-secret (owned by broker CR)
+	secretKey := types.NamespacedName{
+		Name:      artemis.Name + "-applied-security",
+		Namespace: artemis.Namespace,
+	}
+	appliedSecSecret := corev1.Secret{}
+	err := r.Client.Get(context.TODO(), secretKey, &appliedSecSecret)
```

Comment:
This looks like we are again adding state to the reconcile loop, I don't think this is wise. The reconcile process should not be conditional on the exist of a secret, this is essentially two sources of truth. What will make them converge and or stop them diverging. I think we have to make it such that the process of applying is fast and idempotent.

Reply:
Yeah that one need some more thought. I'm still trying to figure that out.